### PR TITLE
Further promisifying and removing async.js

### PR DESCRIPTION
**Pull request by codingtwinky**
I really hate async.js.  It's literally a putting a lipstick on a pig named callback.  Look at the [note](https://caolan.github.io/async/docs.html#parallel), why the hell is it named ".parallel"?  Maybe I'm too strongly opinionated about this.

But thankfully, the world of javascript moved on to promises and coroutines and we can get rid of this relic of the past.

### CODE REVIEW

File: test/spec.git-api.conflict.js [hunk 1]
```diff
@@ -26,21 +24,18 @@ describe('git-api conflict rebase', function () {
...
-				function(done) { common.post(req, '/testing/changefile', { file: path.join(testDir, testFile1) }, done); },
-				function(done) { common.post(req, '/commit', { path: testDir, message: commitMessage, files: [{ name: testFile1 }] }, done); }
-			], done);
-		});
+			return common.post(req, '/testing/createfile', { file: path.join(testDir, testFile1) })
+        .then(function() { return common.post(req, '/commit', { path: testDir, message: commitMessage, files: [{ name: testFile1 }] }); })
+        .then(function() { return common.post(req, '/branches', { path: testDir, name: testBranch, startPoint: 'master'}); })
+        .then(function() { return common.post(req, '/testing/changefile', { file: path.join(testDir, testFile1)}); })
+        .then(function() { return common.post(req, '/commit', { path: testDir, message: commitMessage, files: [{ name: testFile1 }]}); })
+				.then(function() { return common.post(req, '/checkout', { path: testDir, name: testBranch}); })
```
Comment (campersau) [inline test/spec.git-api.conflict.js]: Tab vs Spaces?

File: bin/credentials-helper [hunk 2]
```diff
@@ -12,12 +11,9 @@ var usageStatistics = require('../src/usage-statistics');
 
 process.on('uncaughtException', function(err) {
   console.error(err.stack.toString());
-  async.parallel([
-    bugtracker.notify.bind(bugtracker, err, 'credentials-helper'),
-    usageStatistics.addEvent.bind(usageStatistics, 'credentials-helper-exception')
-  ], function() {
-    process.exit();
-  });
+  bugtracker.notify(err, 'credentials-helper');
```
Comment (FredrikNoren) [inline bin/credentials-helper]: This code isn't equivalent; the old code waits for bugtracker and usage statistics to finish before calling process.exit (which immediately terminates the process and doesn't wait for remaining tasks afaik). Not terribly important but want to make sure the implications of the change are known

File: CHANGELOG.md [hunk 3]
```diff
@@ -3,15 +3,20 @@ All notable changes to this project will be documented in this file.
...
+- 1.1.12:
+    - Refactor and promisify clicktests [#879](https://github.com/FredrikNoren/ungit/pull/879)
 - 1.1.11:
     - Fix cli arguments [#871](https://github.com/FredrikNoren/ungit/pull/871)
     - Stop if ~/.ungitrc contains syntax error
     - Removed official support ini format of ~/.ungitrc, because internal API supports only JSON
-- 1.1.10: Fix broken diff out in some cases when diff contains table. [#881](https://github.com/FredrikNoren/ungit/pull/881)
+- 1.1.10:
+    - Fix broken diff out in some cases when diff contains table. [#881](https://github.com/FredrikNoren/ungit/pull/881)
+    - Refactor to remove async and using more promises. [#878](https://github.com/FredrikNoren/ungit/pull/878)
```
Comment (FredrikNoren) [inline CHANGELOG.md]: I'm confused, is this branch based on something else that isn't merged or why are there changes in the _past_ in the changelog here?

Comment (codingtwinky) [inline CHANGELOG.md]: There were some releases since I have created this PR and made that changelog update, I will have to fix this.

**Comment by codingtwinky** [comment]
That is not at all controversial request at all and I admit this PR is gotten out of hands.  This PR is an example of "while I'm at it..." and "oh it would be more efficient if..." along with me increasing scope.  

Originally was intended to simply get rid of `async.js` and I ended up refactoring directory watch code and tried to fix existing intermittent test failures.  

I'm going to keep this open and start creating smaller PRs based on this PR but I think end code will look similar to this one.

**Comment by FredrikNoren** [comment]
I've been playing around with https://github.com/segmentio/nightmare as a replacement for phantom in another project. It's looking really promising, and I would suspect that they support async/await as it's based on electron which is running a more modern chrome version.

**Pull request closed without merging**